src/btree/bt_read.c

```diff
@@ -184,6 +184,281 @@ __bt_reconstruct_delta(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *delta)
+    }
+    for (; j < delta_entries; j++)
+        if (!F_ISSET(delta[j], WT_DELTA_INT_IS_DELETE))
+            __bt_build_ref(session, ref, NULL, NULL, delta[j], false, &refs[final_entries++], incr);
+
+    WT_ASSERT(session, final_entries != 0);
+    WT_ASSERT(session, final_entries < estimated_entries && refs[final_entries] == NULL);
+
+    *ref_entriesp = final_entries;
+
+err:
+    __wt_free(session, base);
+    return (ret);
+}
+
+/*
+ * __bt_reconstruct_internal_deltas --
+ *     Reconstruct delta on an internal page
+ */
+static int
+__bt_reconstruct_internal_deltas(
+  WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *deltas, size_t delta_size)
+{
+    WT_CELL_UNPACK_DELTA_INT **unpacked_deltas, **unpacked_deltas_merged;
+    WT_DECL_RET;
+    WT_DELTA_HEADER *header;
+    WT_PAGE_INDEX *pindex;
+    WT_REF **refs;
+    uint32_t i, j;
+    size_t *delta_size_each, incr, pindex_size, refs_entries, unpacked_deltas_merged_size;
+
+    unpacked_deltas = unpacked_deltas_merged = NULL;
+    refs = NULL;
+    pindex = NULL;
+    unpacked_deltas_merged_size = refs_entries = incr = 0;
+
+    WT_RET(__wt_calloc_def(session, delta_size, &delta_size_each));
+    WT_ERR(__wt_calloc_def(session, delta_size, &unpacked_deltas));
+    for (i = 0, j = 0; i < (uint32_t) delta_size; ++i, j = 0) {
+        header = (WT_DELTA_HEADER *)deltas[i].data;
+        WT_ASSERT(session, header->u.entries != 0);
+        delta_size_each[i] = (size_t)header->u.entries;
+
+        WT_ERR(__wt_calloc_def(session, header->u.entries, &unpacked_deltas[i]));
+        WT_CELL_FOREACH_DELTA_INT(session, ref->page->dsk, header, unpacked_deltas[i][j])
+        {
+            j++;
+        }
+        WT_CELL_FOREACH_END;
+    }
+
+    WT_ERR(__bt_merge_internal_deltas(session, unpacked_deltas, 0, (uint32_t)delta_size - 1,
+      delta_size_each, &unpacked_deltas_merged, &unpacked_deltas_merged_size));
+
+    WT_ERR(__bt_merge_internal_delta_with_base_image(session, ref, unpacked_deltas_merged,
+      unpacked_deltas_merged_size, &refs, &refs_entries, &incr));
+
+    pindex_size = sizeof(WT_PAGE_INDEX) + refs_entries * sizeof(WT_REF *);
+    WT_ERR(__wt_calloc(session, 1, pindex_size, &pindex));
+    incr += pindex_size;
+    pindex->index = (WT_REF **)(pindex + 1);
+    pindex->entries = (uint32_t) refs_entries;
+
+    for (i = 0; i <  pindex->entries; ++i) {
+        refs[i]->pindex_hint = i;
+        pindex->index[i] = refs[i];
+    }
+    pindex->entries = (uint32_t) refs_entries;
+    WT_INTL_INDEX_SET(ref->page, pindex);
+    __wt_cache_page_inmem_incr(session, ref->page, incr);
+
+    if (0) {
+err:
+        if (refs != NULL)
+            for (i = 0; i < (uint32_t) refs_entries; ++i)
+                __wt_free(session, refs[i]);
+    }
+    if (unpacked_deltas != NULL) {
+        for (i = 0; i < (uint32_t) delta_size; ++i)
+            __wt_free(session, unpacked_deltas[i]);
+        __wt_free(session, unpacked_deltas);
+    }
+    __wt_free(session, refs);
```

Comment:
This is also temporary?

Reply:
yes, the refs array holding ref pointers is temporary as the refs have been moved to the pindex.